File: Week 1/cpp/16.operators.cpp
/*
 * WEEK 1 - C++ FUNDAMENTALS
 * Topic: Operators in C++ (Arithmetic, Relational, Logical, Bitwise)
 * File: 16.operators.cpp
 *
 * CONCEPT:
 * C++ exposes essentially the same operator set as Java, with a couple of
 * extras: pointer arithmetic (`*`, `->`, `&`), the scope resolution operator
 * (`::`), the comma operator, and ternary `? :`. The keywords `and`, `or`,
 * `not` exist as alternative spellings of `&&`, `||`, `!`.
 *
 * KEY POINTS:
 *  - Arithmetic: + - * / %       (no `**`; use std::pow)
 *  - Relational: == != > < >= <= -> bool
 *  - Logical:    && || !          (or `and` / `or` / `not`)
 *  - Bitwise:    & | ^ ~ << >>   (no `>>>`)
 *  - Increment/Decrement: ++ -- (prefix and postfix variants)
 *  - Assignment: = += -= *= /= %= &= |= ^= <<= >>=
 *  - Conditional: condition ? a : b
 *  - Comma: e1, e2  -> evaluates both, returns e2
 *
 * SYNTAX:
 *   bool t = (a > b) && (c < d);
 *   int  m = (a > b) ? a : b;
 *   int  q = a / b, r = a % b;
 *
 * DRY RUN:
 *  19 & 28 -> 16; 19 | 28 -> 31; 19 ^ 28 -> 15
 *  ~19 -> -20; 19 << 2 -> 76; 19 >> 2 -> 4
 *  (5 > 6) -> false; ((5>4) && (4>3)) -> true
 */

#include <iostream>
#include <bitset>

int main() {
    std::cout << "=== Arithmetic ===\n";
    std::cout << "10 + 3 = " << (10 + 3) << '\n';
    std::cout << "10 / 3 = " << (10 / 3) << "  (int truncation)\n";
    std::cout << "10 % 3 = " << (10 % 3) << '\n';
    std::cout << "10.0/3 = " << (10.0 / 3) << "  (double)\n";

    std::cout << "\n=== Relational ===\n";
    int a = 5, b = 6;
    std::cout << std::boolalpha;        // print bools as 'true'/'false'
    std::cout << "a == b: " << (a == b) << '\n';
    std::cout << "a != b: " << (a != b) << '\n';
    std::cout << "a >  b: " << (a >  b) << '\n';
    std::cout << "a <  b: " << (a <  b) << '\n';

    std::cout << "\n=== Logical ===\n";
    bool x = true, y = false;
    std::cout << "x && y: " << (x && y) << '\n';
    std::cout << "x || y: " << (x || y) << '\n';
    std::cout << "!x    : " << (!x) << '\n';
    std::cout << "x and y (alt spelling): " << (x and y) << '\n';

    std::cout << "\n=== Bitwise ===\n";
    int p = 19, q = 28;             // 19 = 10011, 28 = 11100
    std::cout << "p     = " << std::bitset<8>(p) << '\n';
    std::cout << "q     = " << std::bitset<8>(q) << '\n';
    std::cout << "p & q = " << (p & q) << "  (== " << std::bitset<8>(p & q) << ")\n";
    std::cout << "p | q = " << (p | q) << '\n';
    std::cout << "p ^ q = " << (p ^ q) << '\n';
    std::cout << "~p    = " << (~p) << '\n';
    std::cout << "p<<2  = " << (p << 2) << '\n';
    std::cout << "p>>2  = " << (p >> 2) << '\n';

    std::cout << "\n=== Conditional & comma ===\n";
    int m = (a > b) ? a : b;
    std::cout << "max(a,b) via ?: = " << m << '\n';
    int n = (a += 1, b += 2, a + b);    // evaluates left-to-right; n = a+b
    std::cout << "comma op result n = " << n << " (a now " << a << ", b now " << b << ")\n";

    return 0;
}

/*
 * NOTES:
 *  - C++ has NO `>>>` (unsigned right shift). For an unsigned shift cast to
 *    `unsigned` (or `uint32_t`) first.
 *  - `&&` / `||` short-circuit, exactly like Java and Python.
 *  - Operator overloading lets you reuse `+`, `<<`, etc. on custom types
 *    (Java forbids operator overloading except for `+` on String).
 *  - `<bitset>` is great for visualising binary values in output.
 */

File: Week 1/python/16.operators.py
"""
WEEK 1 - PYTHON FUNDAMENTALS
Topic: Operators in Python (Arithmetic, Relational, Logical, Bitwise)
File: 16.operators.py

CONCEPT:
Python supports the same operator categories as Java with a few notable
differences. The biggest changes are:
  - Logical operators are SPELLED OUT: `and`, `or`, `not` (Java uses `&&`, `||`, `!`)
  - `/` is true division; use `//` for floor division
  - `**` is exponentiation
  - Chained comparisons are supported: `a < b < c`

KEY POINTS:
- Arithmetic: + - * / // % **
- Relational: == != > < >= <=
- Logical:    and or not (with short-circuit evaluation)
- Bitwise:    & | ^ ~ << >>
- Identity:   is, is not  (compare object identity, NOT equality)
- Membership: in, not in  (`x in [1,2,3]`)
- Assignment: = += -= *= /= //= %= **= &= |= ^= <<= >>=

PYTHON-SPECIFIC NOTES vs Java:
- `==` compares VALUES; `is` compares IDENTITY (Java's `==` is identity for objects).
- `and` / `or` return the OPERAND VALUE, not a boolean ("hi" or "" -> "hi").
- Chained comparisons: `1 < x < 10` is mathematically what you'd expect (Java
  forbids this — would parse as (1<x)<10 with type errors).
- No unsigned-right-shift (`>>>`) — Python ints have no fixed width.

DRY RUN:
1. a, b = True, False
   a and b -> False
   a or  b -> True
   not a    -> False
2. 19 & 28 -> 16; 19 | 28 -> 31; 19 ^ 28 -> 15
3. 1 < 5 < 10 -> True (chained)
"""


def show_arithmetic() -> None:
    print("=== Arithmetic ===")
    print(f"  10 + 3 = {10 + 3}")
    print(f"  10 - 3 = {10 - 3}")
    print(f"  10 * 3 = {10 * 3}")
    print(f"  10 / 3 = {10 / 3}      # true division -> float")
    print(f"  10 // 3 = {10 // 3}      # floor division -> int")
    print(f"  10 % 3 = {10 % 3}")
    print(f"  2 ** 10 = {2 ** 10}    # exponentiation")


def show_relational() -> None:
    print("\n=== Relational (each returns bool) ===")
    a, b = 5, 6
    print(f"  a == b: {a == b}")
    print(f"  a != b: {a != b}")
    print(f"  a >  b: {a > b}")
    print(f"  a <  b: {a < b}")
    print(f"  a >= b: {a >= b}")
    print(f"  a <= b: {a <= b}")
    # Chained comparisons — illegal in Java
    x = 7
    print(f"  1 < {x} < 10: {1 < x < 10}    # chained, evaluates as (1<x) and (x<10)")


def show_logical() -> None:
    print("\n=== Logical (and / or / not) ===")
    a, b = True, False
    print(f"  a and b: {a and b}")
    print(f"  a or  b: {a or b}")
    print(f"  not a  : {not a}")
    # Short-circuit value pass-through
    print(f"  '' or 'fallback': {'' or 'fallback'!r}    # returns first truthy operand")
    print(f"  'hi' and 42: {'hi' and 42!r}              # returns last operand if all truthy")


def show_bitwise() -> None:
    print("\n=== Bitwise ===")
    a, b = 19, 28          # 19 = 10011, 28 = 11100
    print(f"  a & b  = {a & b}    # AND -> 16")
    print(f"  a | b  = {a | b}    # OR  -> 31")
    print(f"  a ^ b  = {a ^ b}    # XOR -> 15")
    print(f"  ~a     = {~a}     # NOT -> -(a+1) = -20")
    print(f"  a << 2 = {a << 2}    # left shift -> 76")
    print(f"  a >> 2 = {a >> 2}    # right shift -> 4")
    # No `>>>` in Python; emulate with mask
    n = -4
    print(f"  unsigned-right-shift of -4 by 28 (32-bit emulation) = {(n & 0xFFFFFFFF) >> 28}")


def show_identity_membership() -> None:
    print("\n=== Identity / Membership ===")
    a = [1, 2, 3]
    b = a
    c = [1, 2, 3]
    print(f"  a is b: {a is b}    # same object -> True")
    print(f"  a is c: {a is c}    # different objects, same value -> False")
    print(f"  a == c: {a == c}    # value comparison -> True")
    print(f"  2 in a: {2 in a}")
    print(f"  9 not in a: {9 not in a}")


def main() -> None:
    show_arithmetic()
    show_relational()
    show_logical()
    show_bitwise()
    show_identity_membership()


if __name__ == "__main__":
    main()


# NOTES:
# - Java's `&&`/`||`/`!` are spelled `and`/`or`/`not` in Python.
# - Java uses `==` for both primitive equality AND reference identity.
#   Python uses `==` for VALUE equality and `is` for IDENTITY.
# - Python lacks `>>>` (unsigned right shift) because ints aren't fixed width.
# - `and` / `or` return one of the operands (not strictly bool); convenient idiom.
# - Chained comparisons like `0 <= x < n` make range checks easy.
